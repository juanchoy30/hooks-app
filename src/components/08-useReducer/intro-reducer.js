
const initialState = [{
    id: 1,
    todo: 'Buy milk',
    done: false
}];

//// Do not use .push() when using reducer (When working on React). It can lead to errors
//// .push() modifes or mutates the object, it React the state is what has to be modifed

const todoReducer = ( state = initialState, action ) => {
    
    if ( action?.type === 'add' ) {           //// ? will let now the function to execute only if action exists
        return [ ...state, action.payload ];
    }
    
    return state;
};

let todos = todoReducer();


const newTodo = {
    id: 2,
    todo: 'Buy bread',
    done: false
};

const addTodoAction = {
    type: 'add',
    payload: newTodo
}


todos = todoReducer( todos, addTodoAction );


console.log(todos);