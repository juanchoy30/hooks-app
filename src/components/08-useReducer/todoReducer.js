
export const todoReducer = ( state = [], action ) => {

    switch ( action.type ) {
        case 'add':
            return [ ...state, action.payload ];

        case 'delete':
            return state.filter( todo => todo.id !== action.payload );

        case 'toggle':                // Short vertion
            return state.map( todo => 
                    ( todo.id === action.payload )
                        ? { ...todo, done: !todo.done }
                        : todo
                )

        case 'toggle-old':            // Long vertion, to understand the logic
            return state.map( todo => {

                if ( todo.id === action.payload ) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }

            })
    
        default:
            return state;
    }

}