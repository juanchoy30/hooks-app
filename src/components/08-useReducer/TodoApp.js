import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './styles.css';


// useState and useReducer are almost the same. However, useState should be used
// when simple states or few things have to be change. When complex or many actions
// have to me run, useReducer must be used.

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];    /// || if null return [] empty array
};

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer( todoReducer, [], init );

    useEffect( ()=> {
        localStorage.setItem('todos', JSON.stringify( todos ));
    }, [todos] );

    const handleDelete = ( todoId ) => {
        console.log(todoId)

        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch( action );
    }

    const handleToggle = ( todoId ) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    } 


    const handleAddTodo = ( newTodo ) => {

        dispatch({
            type: 'add',
            payload: newTodo
        });

    }


    return (
        <div>
            <h1>Todo App <small>( { todos.length } )</small></h1>
            <hr />

            <div className="row">

                <div className="col-7">

                    <TodoList 
                        todos={ todos }
                        handleToggle={ handleToggle } 
                        handleDelete={ handleDelete }
                    />

                </div>

                <div className="col-5">

                    <TodoAdd 
                        handleAddTodo={ handleAddTodo }
                    />

                </div>

            </div>


        </div>
    )
}
