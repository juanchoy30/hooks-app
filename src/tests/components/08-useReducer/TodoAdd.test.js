import React from 'react';
import { shallow } from 'enzyme';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('tests on <TodoAdd />', () => {

    const handleAddTodo = jest.fn();

    const wrapper = shallow(
        <TodoAdd 
            handleAddTodo={ handleAddTodo }
        />
    )

    test('should render correctly', () => {

        expect( wrapper ).toMatchSnapshot();
        
    });

    test('should not call handleAddTodo', () => {
       
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} });   // The event must be sent as argument because of the preventDefault() // preventDefault(){} short form to indicate function
        expect( handleAddTodo ).toHaveBeenCalledTimes(0);
    
    });
    
    test('should call handleAddTodo', () => { 

        const value = 'Aprender React';
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} }); 
        expect( handleAddTodo ).toHaveBeenCalledTimes(1);
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) ); //This test could pass with an empty object {}, so it has to be settled more precisely
        
        expect( handleAddTodo ).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc:value,
            done:false
        });

        expect( wrapper.find('input').prop('value') ).toBe('');

    });

})
