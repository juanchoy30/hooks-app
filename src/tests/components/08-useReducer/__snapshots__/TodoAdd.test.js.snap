// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <TodoAdd /> should render correctly 1`] = `
<Fragment>
  <h4>
    Add todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Learn..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary btn-block mt-1"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
